File: apps/api/src/routes/syntax.ts
import { Router } from "express";
import { and, eq } from "drizzle-orm";
import {
  db,
  syntaxLessonsTable,
  languagesTable,
} from "@devlearn/database";
import { asyncHandler } from "../lib/async.ts";
import { HttpError } from "../lib/errors.ts";
import { parseIntParam, parseOptionalInt, parseOptionalString } from "../lib/parse.ts";

export const syntaxRouter = Router();

const syntaxSelect = {
  id: syntaxLessonsTable.id,
  languageId: syntaxLessonsTable.languageId,
  title: syntaxLessonsTable.title,
  concept: syntaxLessonsTable.concept,
  rawSyntax: syntaxLessonsTable.rawSyntax,
  explanation: syntaxLessonsTable.explanation,
  realWorldExample: syntaxLessonsTable.realWorldExample,
  githubProject: syntaxLessonsTable.githubProject,
  githubUrl: syntaxLessonsTable.githubUrl,
  difficulty: syntaxLessonsTable.difficulty,
  category: syntaxLessonsTable.category,
  orderIndex: syntaxLessonsTable.orderIndex,
  languageName: languagesTable.name,
};

syntaxRouter.get(
  "/syntax-lessons",
  asyncHandler(async (req, res) => {
    const languageId = parseOptionalInt(req.query.languageId);
    const difficulty = parseOptionalString(req.query.difficulty);
    const category = parseOptionalString(req.query.category);

    const where = [];
    if (languageId !== undefined) where.push(eq(syntaxLessonsTable.languageId, languageId));
    if (difficulty) where.push(eq(syntaxLessonsTable.difficulty, difficulty));
    if (category) where.push(eq(syntaxLessonsTable.category, category));

    const rows = await db
      .select(syntaxSelect)
      .from(syntaxLessonsTable)
      .leftJoin(languagesTable, eq(languagesTable.id, syntaxLessonsTable.languageId))
      .where(where.length ? and(...where) : undefined)
      .orderBy(syntaxLessonsTable.orderIndex, syntaxLessonsTable.id);

    res.json(rows);
  }),
);

syntaxRouter.get(
  "/syntax-lessons/:id",
  asyncHandler(async (req, res) => {
    const id = parseIntParam(req.params.id, "id");
    const [row] = await db
      .select(syntaxSelect)
      .from(syntaxLessonsTable)
      .leftJoin(languagesTable, eq(languagesTable.id, syntaxLessonsTable.languageId))
      .where(eq(syntaxLessonsTable.id, id));
    if (!row) throw new HttpError(404, "Syntax lesson not found");
    res.json(row);
  }),
);

File: apps/api/app/routers/syntax.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db import get_db
from app.models import Language, SyntaxLesson
from app.schemas import SyntaxLessonOut

router = APIRouter(tags=["syntax"])


@router.get("/syntax-lessons", response_model=list[SyntaxLessonOut])
def list_syntax_lessons(
    language_id: int | None = Query(None, alias="languageId"),
    difficulty: str | None = None,
    category: str | None = None,
    db: Session = Depends(get_db),
) -> list[SyntaxLessonOut]:
    stmt = (
        select(
            SyntaxLesson.id,
            SyntaxLesson.language_id,
            SyntaxLesson.title,
            SyntaxLesson.concept,
            SyntaxLesson.raw_syntax,
            SyntaxLesson.explanation,
            SyntaxLesson.real_world_example,
            SyntaxLesson.github_project,
            SyntaxLesson.github_url,
            SyntaxLesson.difficulty,
            SyntaxLesson.category,
            SyntaxLesson.order_index,
            Language.name.label("language_name"),
        )
        .outerjoin(Language, Language.id == SyntaxLesson.language_id)
        .order_by(SyntaxLesson.order_index, SyntaxLesson.id)
    )
    if language_id is not None:
        stmt = stmt.where(SyntaxLesson.language_id == language_id)
    if difficulty:
        stmt = stmt.where(SyntaxLesson.difficulty == difficulty)
    if category:
        stmt = stmt.where(SyntaxLesson.category == category)
    rows = db.execute(stmt).mappings().all()
    return [SyntaxLessonOut.model_validate(dict(r)) for r in rows]


@router.get("/syntax-lessons/{lesson_id}", response_model=SyntaxLessonOut)
def get_syntax_lesson(lesson_id: int, db: Session = Depends(get_db)) -> SyntaxLessonOut:
    stmt = (
        select(
            SyntaxLesson.id,
            SyntaxLesson.language_id,
            SyntaxLesson.title,
            SyntaxLesson.concept,
            SyntaxLesson.raw_syntax,
            SyntaxLesson.explanation,
            SyntaxLesson.real_world_example,
            SyntaxLesson.github_project,
            SyntaxLesson.github_url,
            SyntaxLesson.difficulty,
            SyntaxLesson.category,
            SyntaxLesson.order_index,
            Language.name.label("language_name"),
        )
        .outerjoin(Language, Language.id == SyntaxLesson.language_id)
        .where(SyntaxLesson.id == lesson_id)
    )
    row = db.execute(stmt).mappings().first()
    if not row:
        raise HTTPException(404, "Syntax lesson not found")
    return SyntaxLessonOut.model_validate(dict(row))
